### PR TITLE
traccar: init at 6.5

### LINKED ISSUE #261879: Package request: traccar-server + traccar-web
**Project description**

Traccar is the leading GPS tracking software. Vehicle and personal tracking. Real time view, reports, notifications.

It's a self hosted server + web app to track devices like phones or other GPS devices.

It uses Maven.

Dockerfiles available and looks like is not that hard!

https://github.com/traccar/traccar-docker/blob/master/Dockerfile.alpine
https://github.com/traccar/traccar-docker/blob/master/Dockerfile.ubuntu
https://github.com/traccar/traccar-docker/blob/master/Dockerfile.debian

**Metadata**

* homepage URL: https://www.traccar.org/ - https://www.traccar.org/server/
* source URL: https://github.com/traccar/traccar - https://github.com/traccar/traccar-web
* license: Apache License 2.0
* platforms: unix, linux, darwin, ...

### PR DESCRIPTION
**Pull request by frederictobiasc**
## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

closes https://github.com/NixOS/nixpkgs/issues/261879

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by KiaraGrouwstra** [comment]
Result of `nixpkgs-review pr 354732` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>traccar</li>
  </ul>
</details>

File: nixos/modules/services/monitoring/traccar.nix [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+    databaseUrl = "jdbc:h2:${stateDirectory}/traccar";
+    databaseUser = "sa";
+    loggerConsole = "true";
+    mediaPath = "${stateDirectory}/media";
+    templatesRoot = "${stateDirectory}/templates";
+  };
+in
+{
+  options.services.traccar = {
+    enable = lib.mkEnableOption "Whether to enable Traccar, an open source GPS tracking system.";
```
Comment (FliegendeWurst) [inline nixos/modules/services/monitoring/traccar.nix:35]:
```suggestion
    enable = lib.mkEnableOption "Traccar, an open source GPS tracking system";
```

File: nixos/modules/services/monitoring/traccar.nix [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+          ${pkgs.envsubst}/bin/envsubst \
+            -i ${configuration} \
+            -o ${configFilePath}
+          umask $old_umask
+        '';
+
+        serviceConfig = {
+          DynamicUser = true;
+          EnvironmentFile = cfg.environmentFile;
+          ExecStart = "${pkgs.openjdk}/bin/java -jar ${pkgs.traccar}/tracker-server.jar ${configFilePath}";
```
Comment (FliegendeWurst) [inline nixos/modules/services/monitoring/traccar.nix:96]: You could have this as a wrapper in the traccar package.

File: pkgs/by-name/tr/traccar/package.nix [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    for dir in lib schema templates web ; do
+      mkdir -p $out/$dir
+      cp -a $dir $out
+    done
+    mkdir $out/bin
+    cp tracker-server.jar $out/tracker-server.jar
+  '';
+
+  meta = with lib; {
+    description = "Traccar, an open source GPS tracking system.";
```
Comment (FliegendeWurst) [inline pkgs/by-name/tr/traccar/package.nix:26]:
```suggestion
    description = "Open source GPS tracking system";
```

File: pkgs/by-name/tr/traccar/package.nix [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+  pname = "traccar";
+  version = "6.5";
+
+  src = fetchzip {
+    stripRoot = false;
+    url = "https://github.com/traccar/traccar/releases/download/v${version}/traccar-other-${version}.zip";
+    hash = "sha256-XCG3G24oe/qR6LiMJASb9STOnyTCtw+2HigaPawcQvU=";
+  };
+
+  installPhase = ''
```
Comment (FliegendeWurst) [inline pkgs/by-name/tr/traccar/package.nix:16]: should include preInstall and postInstall

**Review by FliegendeWurst — CHANGES_REQUESTED** [review]

File: nixos/modules/services/monitoring/traccar.nix (hunk 1 above)
Comment (frederictobiasc) [reply nixos/modules/services/monitoring/traccar.nix:35]: :+1:

File: nixos/modules/services/monitoring/traccar.nix (hunk 2 above)
Comment (frederictobiasc) [reply nixos/modules/services/monitoring/traccar.nix:96]: :+1:

File: pkgs/by-name/tr/traccar/package.nix (hunk 4 above)
Comment (frederictobiasc) [reply pkgs/by-name/tr/traccar/package.nix:16]: :+1:

File: pkgs/by-name/tr/traccar/package.nix (hunk 3 above)
Comment (frederictobiasc) [reply pkgs/by-name/tr/traccar/package.nix:26]: :+1:

**Review by frederictobiasc — COMMENTED** [review]
Thanks for your review!

File: nixos/modules/services/monitoring/traccar.nix [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+          ${pkgs.envsubst}/bin/envsubst \
+            -i ${configuration} \
+            -o ${configFilePath}
+          umask $old_umask
+        '';
+
+        serviceConfig = {
+          DynamicUser = true;
+          EnvironmentFile = cfg.environmentFile;
+          ExecStart = "${pkgs.traccar}/traccar-wrapped ${configFilePath}";
```
Comment (FliegendeWurst) [inline nixos/modules/services/monitoring/traccar.nix:96]:
```suggestion
          ExecStart = "${lib.getExe pkgs.traccar} ${configFilePath}";
```

File: pkgs/by-name/tr/traccar/package.nix [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+    for dir in lib schema templates web ; do
+      mkdir -p $out/$dir
+      cp -a $dir $out
+    done
+
+    mkdir -p $out/share/traccar
+    install -Dm644 tracker-server.jar $out
+
+    makeWrapper ${pkgs.openjdk}/bin/java $out/traccar-wrapped \
```
Comment (FliegendeWurst) [inline pkgs/by-name/tr/traccar/package.nix:29]:
```suggestion
    makeWrapper ${pkgs.openjdk}/bin/java $out/bin/traccar \
```

File: pkgs/by-name/tr/traccar/package.nix [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+    runHook postInstall
+  '';
+
+  meta = with lib; {
+    description = "Open source GPS tracking system";
+    homepage = "https://www.traccar.org/";
+    sourceProvenance = with sourceTypes; [ binaryBytecode ];
+    license = licenses.asl20;
+    maintainers = with maintainers; [ frederictobiasc ];
```
Comment (FliegendeWurst) [inline pkgs/by-name/tr/traccar/package.nix:40]:
```suggestion
    maintainers = with maintainers; [ frederictobiasc ];
    mainProgram = "traccar";
```

**Comment by frederictobiasc** [comment]
@FliegendeWurst thanks again for your review! I applied all your suggestions and tested it successfully.

**Review by FliegendeWurst — APPROVED** [review]

**Review by FliegendeWurst — APPROVED** [review]

File: nixos/modules/services/monitoring/traccar.nix [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+        preStart = ''
+          # Copy new templates into our state directory.
+          cp -a --update=none ${pkgs.traccar}/templates ${stateDirectory}
+          test -f '${configFilePath}' && rm -f '${configFilePath}'
+
+          # Substitute the configFile from Envvars read from EnvironmentFile
+          old_umask=$(umask)
+          umask 0177
+          ${lib.getExe pkgs.envsubst}/bin/envsubst \
```
Comment (FliegendeWurst) [inline nixos/modules/services/monitoring/traccar.nix:87]:
```suggestion
          ${lib.getExe pkgs.envsubst} \
```

**Review by FliegendeWurst — APPROVED** [review]

**Pull request merged**